contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +895,85 @@ clearly described and the permissions issues should be resolved.
 This could be done with a double opt-in (an opt-in from both the referrer and the refer-ee) or with secondary permissions
 checks performed in admission. 
 
-TODO: Plugins, extensions, nested kinds, headers
+### References to a single resource type 
+
+A single kind object reference is straightforward in that the controller can hard-code most qualifiers needed to identify the object. As such as the only value needed to be provided is the name (and namespace, although cross-namespace references are discouraged):
+
+```yaml
+# for a single resource, the suffix should be Ref, with the field name
+# providing an indication as to the resource type referenced.
+secretRef: 
+    name: foo
+    # namespace would generally not be needed and is discouraged, 
+    # as explained above.
+    namespace: foo-namespace
+```
+
+#### Controller behavior
+
+The operator is expected to know the version, group, and resource name of the object it needs to retrieve the value from, and can use the discovery client or construct the API path directly.
+
+### References which can point to multiple resource types
+
+Multi-kind object references are used when there is a bounded set of valid resource types that a reference can point to.
+
+As with a single-kind object reference, the operator can supply missing fields, provided that the fields that are present are sufficient to uniquely identify the object resource type among the set of supported types.
+
+```yaml
+# guidance for the field name is the same as a single resource.
+fooRef:
```

Comment:
Just to clarify - the concrete results of this KEP is that we would strongly recommend that *all* new object-reference-ish fields:
1) adopt the ObjectRef style (a `struct` as opposed to `FooName string`); and
2) adopt the `somethingRef` name style (as opposed to other pseudo-conventions)

Yes?

Reply:
We should make it clear that the "foo" or "something" isn't about the thing being referenced (since we'll have `kind` for that), which means it is about what exactly? The use of the referenced thing?

Reply:
Yes, though sometimes that's the same word, especially if the referent is a fixed type.

Reply:
> Yes, though sometimes that's the same word, especially if the referent is a fixed type.

+1. I think I'll break off another section specifically on the naming rationale of the "foo" part of "fooRef", to see if we can make it clear.

Reply:
> Just to clarify - the concrete results of this KEP is that we would strongly recommend that _all_ new object-reference-ish fields:
> 
> 1. adopt the ObjectRef style (a `struct` as opposed to `FooName string`); and
> 2. adopt the `somethingRef` name style (as opposed to other pseudo-conventions)
> 
> Yes?

And yes, that's an accurate list. Although we are figuring out the fieldpath syntax, which may introduce a third change (since previously I presume fieldrefs used javascript field syntax).

Reply:
Added a section to clarify best practices on naming. [PTAL](https://github.com/kubernetes/community/pull/5748/files#diff-54ad8eef5360e2f07b362fa109bc82f60bf9ec318130b56a4735a7403326d9e2R898).

Reply:
New section LGTM, thanks.  I've convinced sig-storage to adopt the fooRef convention already :)

Reply:
awesome!